### DISCUSSION TITLE
How do I fix "Unwanted system library"

### DISCUSSION THREAD

**Discussion by porttack** [Homebrew/discussions#870] [Linux]
I'm trying to use github workfows to build formulas and deploy bottles to github. By default, this builds things on both macos and ubuntu. While I could turn off ubuntu, I'd rather not and am getting an error on my ubuntu 'brew linkage --test' phase:

```
brew linkage --test yogieric/test2/khan-linter
==> FAILED
Unwanted system libraries:
  /usr/lib/x86_64-linux-gnu/libyaml-0.so.2
```

It is actually fired by the `brew test-bot --only-formulae`. Is there anyway to not treat this as a failure or a way I can fix things so this doesn't happen? (I tried adding libyaml as a dependency, but nothing changes.)

**Comment by carlocab** [comment]
Your build is picking up linkage with the system `libyaml`. You need to find a way to stop that (e.g. with a `configure` flag like `--disable-libyaml` or `--without-libyaml`), or add a `libyaml` dependency and find a way to get it to link with brewed `libyaml` (e.g. something like `--with-libyaml=/path/to/brewed/libyaml`).

**Reply by danielnachun** [reply]
I think this could actually be a more complicated problem than just adding `libyaml` as a dependency.  `khan-linter` appears to be a NodeJS package, and in our Linux bottling tracking issue https://github.com/Homebrew/linuxbrew-core/issues/22225 we have two NodeJS packages `appium` and `serverless` which have the same linkage failure as described above.   

As I noted in that issue, the problem is that `npm` by default will download and install precompiled binaries to the formula cellar.  This problem is a great example of why we don't allow formulae to install precompiled binaries in the cellar.  I think it evaded our detection on macOS because those binaries are only using system libraries, and `npm` makes sure to install versions of the binary that will link to the correct versions based on the OS version (i.e. 10.15 vs 11.0).

The solution here should be to force `npm` to rebuild the native binaries when it installs into the cellar, but I haven't been able to figure that out yet.  If anyone with more `npm` expertise knows how to achieve this, that would be great as it would solve a problem that is affecting multiple formulae.

**Reply by Bo98** [reply]
What does `brew linkage --reverse appium` (and `serverless`) return?

**Reply by danielnachun** [reply]
For `appium` I get the following:

```
root@b60a1578876d:/home/linuxbrew# brew linkage --test appium
Missing libraries:
  unexpected (libnspr4.so)
  unexpected (libnss3.so)
  unexpected (libnssutil3.so)
  unexpected (libxcb.so.1)
Unwanted system libraries:
  /lib/x86_64-linux-gnu/libglib-2.0.so.0
```

```
root@b60a1578876d:/home/linuxbrew# brew linkage --reverse appium
/lib/x86_64-linux-gnu/libc.so.6
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

/lib/x86_64-linux-gnu/libdl.so.2
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

/lib/x86_64-linux-gnu/libgcc_s.so.1
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

/lib/x86_64-linux-gnu/libglib-2.0.so.0
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

/lib/x86_64-linux-gnu/libm.so.6
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

/lib/x86_64-linux-gnu/libpthread.so.0
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

/lib/x86_64-linux-gnu/librt.so.1
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

libnspr4.so
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

libnss3.so
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

libnssutil3.so
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

libxcb.so.1
  libexec/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64

**Reply by danielnachun** [reply]
For `serverless` I get the following:

```
root@b60a1578876d:/home/linuxbrew# brew linkage --test serverless
Missing libraries:
  unexpected (libpython3.7m.so.1.0)
```

```
root@b60a1578876d:/home/linuxbrew# brew linkage --reverse serverless
/lib/x86_64-linux-gnu/libc.so.6
  libexec/lib/node_modules/serverless/node_modules/@serverless/enterprise-plugin/sdk-py/serverless_sdk/vendor/wrapt/_wrappers.cpython-37m-x86_64-linux-gnu.so
  libexec/lib/node_modules/serverless/node_modules/snappy/build/Release/binding.node
  libexec/lib/node_modules/serverless/node_modules/snappy/build/Release/obj.target/binding.node

/lib/x86_64-linux-gnu/libpthread.so.0
  libexec/lib/node_modules/serverless/node_modules/@serverless/enterprise-plugin/sdk-py/serverless_sdk/vendor/wrapt/_wrappers.cpython-37m-x86_64-linux-gnu.so
  libexec/lib/node_modules/serverless/node_modules/snappy/build/Release/binding.node
  libexec/lib/node_modules/serverless/node_modules/snappy/build/Release/obj.target/binding.node

/usr/lib/x86_64-linux-gnu/libstdc++.so.6
  libexec/lib/node_modules/serverless/node_modules/snappy/build/Release/binding.node
  libexec/lib/node_modules/serverless/node_modules/snappy/build/Release/obj.target/binding.node

libpython3.7m.so.1.0
  libexec/lib/node_modules/serverless/node_modules/@serverless/enterprise-plugin/sdk-py/serverless_sdk/vendor/wrapt/_wrappers.cpython-37m-x86_64-linux-gnu.so
```

**Reply by Bo98** [reply]
Unfortunately there's nothing we can do globally - it's down to specifics for each module.

For `appium-chromedriver`, you perhaps could skip the binary installation (https://github.com/appium/appium-chromedriver#skipping-binary-installation) and in place put a symlink to the binary in the `chromedriver` formula (if that works - it's a very specific feature in `appium` to test).

For `serverless`, this is a vendored file in `serverless/enterprise-plugin`: https://github.com/serverless/enterprise-plugin/blob/1417c00db392488b00d0e9772b9f4af202d62ab2/sdk-py/serverless_sdk/vendor/wrapt/_wrappers.cpython-37m-x86_64-linux-gnu.so. It can be removed safely as there's a Python fallback (the native version was probably not working anyway).

**Reply by danielnachun** [reply]
Wow, thanks for figuring that out!  @yogieric if you can post the output of `brew linkage --reverse yogieric/test2/khan-linter`, we should be able to figure out which `npm` module is causing the linkage failure, and hopefully figure out how to fix it like we did for these other two formulae.

For `serverless`, I'll just make sure to delete the vendored file.  For `appium`, it looks like `chromedriver` was actually moved from homebrew/core into homebrew/cask.  I'll have to look into why that was the case.  If it's possible to bring it back as a formula, then it can be used as you described.  If not, then I'll just disable `chromedriver` for `appium`.

**Reply by Bo98** [reply]
I've been through the pain before with `joplin`: https://github.com/Homebrew/homebrew-core/pull/52197

Node modules can be very sneaky and it can be hard to spot.

Though I suppose any build script can be like that. Node projects just usually have millions of dependencies.
